libnetwork/osl/interface_linux.go

```diff
@@ -276,22 +323,246 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+		return false, fmt.Errorf("failed to subscribe to link updates: %w", err)
+	}
+
+	// When done (interface upped, or timeout), stop the LinkSubscribe and drain
+	// the result channel. If the result channel isn't closed after a timeout,
+	// log a warning to note the goroutine leak.
+	defer func() {
+		close(upped)
+		drainTimerC := time.After(3 * time.Second)
+		for {
+			select {
+			case _, ok := <-update:
+				if !ok {
+					return
+				}
+			case <-drainTimerC:
+				log.G(ctx).Warn("timeout while waiting for LinkSubscribe to terminate")
+			}
+		}
+	}()
+
+	timerC := time.After(5 * time.Second)
+	for {
+		select {
+		case <-timerC:
+			return false, nil
+		case u, ok := <-update:
+			if !ok {
+				// The netlink package failed to read from its netlink socket. It will
+				// already have called the ErrorCallback, so the issue has been logged.
+				return false, nil
+			}
+			if u.Attrs().Index != ifIndex {
+				continue
+			}
+			log.G(ctx).WithFields(log.Fields{
+				"iface": u.Attrs().Name,
+				"ifi":   u.Attrs().Index,
+				"flags": deviceFlags(u.Flags),
+			}).Debug("link update")
+			if u.Flags&unix.IFF_UP == unix.IFF_UP {
+				return true, nil
+			}
+		}
+	}
+}
+
+// advertiseAddrs triggers send unsolicited ARP and Neighbour Advertisement
+// messages, so that caches are updated with the MAC address currently associated
+// with the interface's IP addresses.
+//
+// IP addresses are recycled quickly when endpoints are dropped on network
+// disconnect or container stop. A new MAC address may have been generated, so
+// this is necessary to avoid packets sent to the old MAC address getting dropped
+// until the ARP/Neighbour cache entries expire.
+//
+// Note that the kernel's arp_notify sysctl setting is not respected.
+func (n *Namespace) advertiseAddrs(ctx context.Context, ifIndex int, i *Interface, nlh nlwrap.Handle) error {
+	mac := i.MacAddress()
+	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(log.Fields{
+		"iface": i.dstName,
+		"ifi":   ifIndex,
+		"mac":   mac.String(),
+		"ip4":   i.Address(),
+		"ip6":   i.AddressIPv6(),
+	}))
+
+	if i.address == nil && i.addressIPv6 == nil {
+		// Nothing to do - for example, a bridge with no configured addresses.
+		log.G(ctx).Debug("No IP addresses to advertise")
+		return nil
+	}
+	if mac == nil {
+		// Nothing to do - for example, a layer-3 ipvlan.
+		log.G(ctx).Debug("No MAC address to advertise")
+		return nil
+	}
+	if i.advertiseAddrNMsgs == 0 {
+		log.G(ctx).Debug("Unsolicited ARP/NA is disabled")
+		return nil
+	}
+
+	arpSender, naSender := n.prepAdvertiseAddrs(ctx, i, ifIndex)
+	if arpSender == nil && naSender == nil {
+		return nil
+	}
+	cleanup := func() {
+		if arpSender != nil {
+			arpSender.Close()
+		}
+		if naSender != nil {
+			naSender.Close()
+		}
+	}
+	stillSending := false
+	defer func() {
+		if !stillSending {
+			cleanup()
+		}
+	}()
+
+	send := func(ctx context.Context, nSent int) error {
```

Comment:
The send function doesn't really care about the `nSent` parameter. It's just there to pass through logging metadata. It smells to me like the advertisement number is nothing more than context for the function call.
```go
send(log.WithLogger(ctx, log.G(ctx).WithFields(log.Fields{"n": nSent + 1})))
```

Reply:
Ok.